### PR TITLE
Support EntityID

### PR DESCRIPTION
**Pull request by Roms1383**
Ok here's support for EntityID, also a good chance for me to get familiar with the new environment.
I had to add an additional include inside `wrapper.h`, at least that's the only way I got `bindgen` to recognize and generate its bindings, but not sure if that's the best practical way to do it.

### CODE REVIEW

File: src/interop.rs [hunk 1]
```diff
@@ -0,0 +1,2 @@
+mod entityid;
```
Comment (jac3km4) [inline src/interop.rs:1]: this file/module shouldn't be named `interop`, the types defined in it wrappers and not C interop types, maybe it should be named `types`

File: src/interop/entityid.rs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+#[repr(transparent)]
+pub struct EntityId(red::ent::EntityID);
+
+impl PartialEq for EntityId {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.hash == other.0.hash
+    }
+}
+
+impl PartialEq<red::ent::EntityID> for EntityId {
```
Comment (jac3km4) [inline src/interop/entityid.rs:16]: `red::ent::EntityID` should remain private and not exposed from the API, so IMO there should be no `impl`s that refer to it

File: src/interop/entityid.rs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    fn eq(&self, other: &red::ent::EntityID) -> bool {
+        self.0.hash.eq(&other.hash)
+    }
+}
+
+impl PartialEq<EntityId> for red::ent::EntityID {
```
Comment (jac3km4) [inline src/interop/entityid.rs:22]: same as above

File: src/interop/entityid.rs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    fn eq(&self, other: &EntityId) -> bool {
+        self.hash.eq(&other.0.hash)
+    }
+}
+
+impl Eq for EntityId {}
+
+impl PartialOrd for EntityId {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.0.hash.partial_cmp(&other.0.hash)
```
Comment (jac3km4) [inline src/interop/entityid.rs:32]: the standard convention in Rust is to delegate PartialOrd to Ord
```suggestion
        Some(self.cmp(other))
```

File: src/interop.rs (hunk 1 above)
Comment (jac3km4) [reply src/interop.rs:1]: it also looks like you forgot to include your modules in `lib.rs`

Comment (jac3km4) [reply src/interop.rs:1]: I think the structure should work like this
`lib.rs` publicly defines the `types` mod
`types.rs` publicly re-exports types from individual files

Comment (jac3km4) [inline src/interop.rs:1]:
```suggestion
mod entity_id;
```

File: src/interop.rs [hunk 5]
```diff
@@ -0,0 +1,2 @@
...
+pub use entityid::*;
```
Comment (jac3km4) [inline src/interop.rs:2]:
```suggestion
pub use entity_id::EntityId;
```

File: src/interop/entityid.rs [hunk 6]
```diff
@@ -0,0 +1,92 @@
+#![allow(dead_code)]
+use std::hash::Hash;
+
+use crate::raw::root::RED4ext as red;
+
+#[derive(Debug, Clone)]
+#[repr(transparent)]
+pub struct EntityId(red::ent::EntityID);
```
Comment (jac3km4) [inline src/interop/entityid.rs:8]: this type is 8 bytes large, we should either derive or implement `Copy`

File: src/interop/entityid.rs [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    pub const fn is_persistable(&self) -> bool {
+        self.0.hash >= red::ent::EntityID_PersistableLowerBound
+            && self.0.hash < red::ent::EntityID_PersistableUpperBound
+    }
+
+    #[inline]
+    pub const fn is_transient(&self) -> bool {
+        self.0.hash != 0 && !self.is_persistable()
+    }
+}
```
Comment (jac3km4) [inline src/interop/entityid.rs:92]: since this is a small Copy-able type, it might be worth to define all receivers as `self` instead of `&self`

**Comment by jac3km4** [comment]
> I had to add an additional include inside wrapper.h, at least that's the only way I got bindgen to recognize and generate its bindings, but not sure if that's the best practical way to do it.

adding includes is the right way to add any types that aren't included by default

**Comment by Roms1383** [comment]
Would it be better to remove this conversion from private inner type? @jac3km4 
```rs
impl From<EntityId> for u64 {
    fn from(EntityId(red::ent::EntityID { hash }): EntityId) -> Self {
        hash
    }
}
```

File: src/interop.rs (hunk 1 above)
Comment (Roms1383) [reply src/interop.rs]: done in f2414cc

File: src/interop/entityid.rs (hunk 2 above)
Comment (Roms1383) [reply src/interop/entityid.rs]: done in 35a9078

File: src/interop/entityid.rs (hunk 3 above)
Comment (Roms1383) [reply src/interop/entityid.rs]: done in 35a9078

File: src/interop/entityid.rs (hunk 4 above)
Comment (Roms1383) [reply src/interop/entityid.rs]: done in 7531b56

File: src/interop.rs (hunk 1 above)
Comment (Roms1383) [reply src/interop.rs]: done in faaebe4

File: src/interop.rs (hunk 5 above)
Comment (Roms1383) [reply src/interop.rs]: done in 9ad325a

File: src/interop/entityid.rs (hunk 6 above)
Comment (Roms1383) [reply src/interop/entityid.rs]: done in 882120e

File: src/interop/entityid.rs (hunk 7 above)
Comment (Roms1383) [reply src/interop/entityid.rs]: done in 8c204a7

**Comment by Roms1383** [comment]
it would be very ergonomic to delegate `std::fmt::Display for EntityID` to its [ToDebugString](https://nativedb.red4ext.com/entEntityID#ToDebugString) but I couldn't find its definition anywhere.

**Comment by jac3km4** [comment]
> Would it be better to remove this conversion from private inner type?
the `impl From<EntityId> for u64` doesn't leak anything internal and seems safe, I think it's fine
> it would be very ergonomic to delegate std::fmt::Display for EntityID to its [ToDebugString](https://nativedb.red4ext.com/entEntityID#ToDebugString) but I couldn't find its definition anywhere.
it could be useful to have a better debug impl but ToDebugString is a redscript function, not red4ext one so it's not in the auto-generated code

**Pull request merged**